core/src/identity/ecdsa.rs

```diff
@@ -249,4 +268,13 @@ mod tests {
         let invalid_msg = "h3ll0 w0rld".as_bytes();
         assert!(!pk.verify(invalid_msg, &sig));
     }
+
+    #[test]
+    fn ecdsa_public_key_eq_implies_hash() {
+        fn prop(SomePublicKey(pub1): SomePublicKey, SomePublicKey(pub2): SomePublicKey) -> bool {
+            pub1 != pub2 || hash(&pub1) == hash(&pub2)
```

Comment:
Is this property correct?

Having two equal pub keys will short-circuit this I believe and we never compare the hash but that is exactly the property we want to check.

I just tried the following and with release mode, that takes about 15 seconds to run. I wasn't patient enough to run it in Debug mode. With 10_000 tests, this fails because it can't generate enough tests to satisfy the `min_tests_passed` property (which defaults to 0).

```rust
    #[test]
    fn ecdsa_public_key_eq_implies_hash() {
        fn prop(SomePublicKey(pub1): SomePublicKey, SomePublicKey(pub2): SomePublicKey) -> TestResult {
            if pub1 != pub2 {
                return TestResult::discard()
            }

            if hash(&pub1) == hash(&pub2) {
                return TestResult::passed()
            } else {
                return TestResult::failed()
            }
        }

        QuickCheck::new()
			.tests(100_000)
            .min_tests_passed(1)
            .quickcheck(prop as fn(_, _) -> _);
    }
```

Thanks for implementing this but I am no longer sure that this is actually the best way to check this. Sorry for leading you down this path.

What if we instead just remove the derived `PartialEq` and implement it manually ourselves by comparing the bytes? That would remove the lint warning, the need for the test and gives us peace of mind that this property actually holds.

Reply:
The property is correct for sure, exactly what clippy and rust docs show https://doc.rust-lang.org/std/hash/trait.Hash.html#hash-and-eq.

No surprise that it's hard to generate 100000 equal keys. If you want to generate and cover equal pairs as thoroughly as unequal pairs, I would tweak the arbitrary implementation (or just write another test that generates one key and clones it). I just did not think it was the goal.

Reply:
> The property is correct for sure, exactly what clippy and rust docs show https://doc.rust-lang.org/std/hash/trait.Hash.html#hash-and-eq.

I thought it through again and it does hold but I have found it hard to understand because it is not expressed the same way as the docs say it :)

The whole point of the property is to check if equal keys have equal hashes. Non equal keys are not interesting so we can discard those test inputs.

Cloning one key and comparing `PartialEq` and `Hash` is not useful by itself. What the lint wants to guard against is that the `PartialEq` impl f.e. uses different fields from the `Hash` impl, thus two differently generated keys could violate the property. Unless you have a non-deterministic Hash impl, the hash of the same key will always be the same so cloning the key doesn't necessarily prove that we are not violating the property.

Like I said above, I don't think a proptest is the right tool here, sorry for leading you down that road earlier.

I'd like to use implement `PartialEq` manually. That will silence the lint and we can see from the impl that we are using the internal byte representation for both.